### PR TITLE
Reimplement per-operator selective build

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39401 Reimplement per-operator selective build**

This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

We don't delete the old selective build as it is currently still
used internally for other reasons that aren't subsumed here.
However, this mechanism can now be used to handle non code generated
registrations IF they are placed within the registration library
that is recompiled per mobile application.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

### CODE REVIEW

File: aten/src/ATen/core/op_registration/op_whitelist.h
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_whitelist.h:8]: yes

Comment (bhosmer) [reply aten/src/ATen/core/op_registration/op_whitelist.h:8]: Might be worth adding to the comment? Don't know if I'm an outlier here in terms of liking that kind of stuff totally spelled out, your call (also LMK if this is jumping the gun on review-type suggestions)

File: c10/util/constexpr_string_functions.h
Comment (bhosmer) [inline c10/util/constexpr_string_functions.h:12]: Quick thought, is there an existing API for this kind of stuff we could emulate, like in Boost or something? (maybe you are already?)

File: torch/library.h
Comment (smessmer) [inline torch/library.h:241]: We should use c10::string_view here and also rewrite the constexpr string functions to use it

**Review by smessmer — DISMISSED** [review]
generally looks very good. Requesting changes because I want to take another look once it's using c10::string_view

**Comment by ezyang** [comment]
I updated the code to use `string_view`, so the whitelist constexpr stuff is shorter (though it smacks of the kind of C-era string manipulation that would be better not to do)

File: c10/util/constexpr_string_functions.h
Comment (bhosmer) [inline c10/util/constexpr_string_functions.h:12]: This is a pretty op-whitelist-specific piece of functionality to be in `c10/util/constexpr_string_functions.h`, no? Why not just put it in `op_whitelist.h`?

OTOH if you're thinking it's the first of a more general-purpose crop of utilities, I think we'd want to rename it to like `semi_delimited_list_contains` (and arguably just put it in `string_view.h`, seems like there's a good amount of this kind of stuff in there already)

**Review by bhosmer — APPROVED** [review]
+1, really nice and clear and well explicated. One minor suggestion inline, and also echoing Sebastian's comment re string_view

File: c10/util/constexpr_string_functions.h
Comment (ezyang) [reply c10/util/constexpr_string_functions.h:12]: There's no good reason; @smessmer's original patch had this in its own header and I've been to lazy to move it.

File: torch/library.h
Comment (ezyang) [reply torch/library.h:241]: Mm, string view here isn't really necessary. Is there any particular reason you want a string view here?

**Comment by ezyang** [comment]
There are some dangerous aspects to this PR.

* Suppose I write `m.impl(TORCH_SELECTIVE_NAME(add), &add)` instead of `TORCH_SELECTIVE_NAME(aten::add)`. Then I will always disable this operator on mobile builds. I think adding a `assert(item.find("::") != string_view::npos);` should help; I'm currently testing to make sure this is the case and will add this to the PR if so.
* The need to replicate namespace for selective build is a bit troublesome. It would be nice to not have to replicate it. However, it seems like it is a bit difficult to do this under the current design. `TORCH_SELECTIVE_NAME` operates without any reference to m, so if you omit the namespace there is no reasonable way to determine what the namespace should be. One way around this is I could define a local constexpr variable like `__the_namespace` and then `TORCH_SELECTIVE_NAME` can tack it onto the operator when it exists, except with the current `TORCH_LIBRARY` macro design I don't have anywhere I can actually define this variable.
* In some situations in the current API, we aren't able to statically tell what the dispatch key / operator name is. For example, if I say `m.impl("aten::add", torch::dispatch(kCPU, add_cpu))`, the CPU key here isn't statically known. This PR will default to including operators in that case. One way to make this a little safer is to add a runtime test and exclude the operator in that case (or raise a warning saying that N is disabled, but we weren't able to prune it from the build). I haven't implemented this yet, looking for some opinions.
* I am wondering if we should reduce the API surface of torch/library.h so that people are only able to specify usage patterns for which we can statically work things out. So for example, remove `torch::dispatch` from the public API surface (and force people to always use `TORCH_LIBRARY_IMPL` instead). Similarly, given the problems with contextual namespaces, maybe we should just force namespaces to be redundantly specified. :/

File: tools/code_analyzer/op_deps_pass.cpp
Comment (ljk53) [inline tools/code_analyzer/op_deps_pass.cpp:752]: lgtm - just to confirm, so contextual namespace and operator namespace will both exist after moving to the new world?

Comment (ezyang) [reply tools/code_analyzer/op_deps_pass.cpp:752]: Yes, although it would be nice if the constexpr machinery could get the contextual namespace so I could get rid of this https://github.com/pytorch/pytorch/pull/39401#issuecomment-639599944

**Comment by ezyang** [comment]
@ljk53 do you know how I would go about debugging the ios failure? It looks legit but doesn't have any information

**Comment by ljk53** [comment]
> @ljk53 do you know how I would go about debugging the ios failure? It looks legit but doesn't have any information

I'd try to run it in xcode. @xta0 do you have some instruction to repro ios failure locally? and/or can we improve the error message to surface more details?

**Comment by ezyang** [comment]
BTW, in fbcode, `//Apps/Internal/PyTorchPlayground:PyTorchPlaygroundTests - PyTorchFBNetTests/testFBNetModel (macos, iphonesimulator)` failed the same way, so I'm betting it's probably the same way (also no real error message)

**Comment by ezyang** [comment]
I pushed an update to just have the implementation bits, but not cut anything over; this unblocks @iseeyuan and also is immediately landable, since previously this was blocked on fbcode internal mobile problems.

File: aten/src/ATen/core/operator_name.h
Comment (smessmer) [inline aten/src/ATen/core/operator_name.h:51]: test cases

File: torch/library.h
Comment (smessmer) [inline torch/library.h:333]: Btw, this is an amazing idea and I somehow had forgotten about it. I was trying to find something that avoids having to type the schema string twice at the call site a couple days ago when talking with Martin about selective build for prim ops but didn't come up with this solution :)

File: torch/library.h
Comment (smessmer) [reply torch/library.h:338]: just generally prefer string_view to const char* because it is more flexible and doesn't have downsides. Say at some point you want to instantiate a SelectiveStr holding only the op name from a `const char*` holding the full schema, `string_view` would easily allow that while `const char*` has to be null terminated. But not really important here.

File: torch/library.h
Comment (smessmer) [inline torch/library.h:481]: This pattern opens up `impl` to be called with any type for `Name` and will yield very bad error messages when called with non-string types or string types that we don't support. Can we keep the old implementation (or maybe actually change it to take string_view instead of const char*) and then explicitly specialize for `SelectiveStr<true>`? The specialization could just forward to the `string_view` version.

**Review by smessmer — APPROVED** [review]

**Pull request closed without merging**
